integration/src/test/java/org/mule/shutdown/HTTPPersistentConnectionsOnShutdownTestCase.java

```diff
@@ -0,0 +1,232 @@
+
+    // Mule isn't stopped yet, since the 3rd connection is still open.
+    assertThat(muleContext.isStopped(), is(false));
+
+    // A second request with the already used connections isn't valid.
+    sendRequest(persistentConnection1, FAST_PROCESSING_ENDPOINT);
+    sendRequest(persistentConnection2, FAST_PROCESSING_ENDPOINT);
+    assertResponse(getResponse(persistentConnection1), false);
+    assertResponse(getResponse(persistentConnection2), false);
+
+    // Mule isn't stopped yet, since the 3rd connection is still open.
+    assertThat(muleContext.isStopped(), is(false));
+
+    // If we close the other connection, mule context stops fast.
+    persistentConnection3.close();
+    waitForStoppedState();
+  }
+
+  private void waitForStoppingState() {
+    new PollingProber(SMALL_POLL_TIMEOUT_MILLIS, POLL_DELAY_MILLIS).check(new Probe() {
+
+      @Override
+      public boolean isSatisfied() {
+        return muleContext.isStopping();
+      }
+
+      @Override
+      public String describeFailure() {
+        return "Timeout waiting for muleContext to be in stopping state";
+      }
+    });
+  }
+
+  private void waitForStoppedState() {
+    new PollingProber(SMALL_POLL_TIMEOUT_MILLIS, POLL_DELAY_MILLIS).check(new Probe() {
+
+      @Override
+      public boolean isSatisfied() {
+        return muleContext.isStopped();
+      }
+
+      @Override
+      public String describeFailure() {
+        return "Timeout waiting for muleContext to be in stopped state";
+      }
+    });
+  }
+
+  private Socket generateIdlePersistentConnection() throws IOException {
+    Socket socket = new Socket("localhost", dynamicPort.getNumber());
+    assertThat(socket.isConnected(), is(true));
+
+    sendRequest(socket, FAST_PROCESSING_ENDPOINT);
+    assertResponse(getResponse(socket), true);
+
+    sendRequest(socket, FAST_PROCESSING_ENDPOINT);
+    assertResponse(getResponse(socket), true);
+
+    return socket;
+  }
+
+  private void sendRequest(Socket socket, String endpoint) throws IOException {
+    PrintWriter writer = new PrintWriter(socket.getOutputStream());
+    writer.println(format("GET %s %s", endpoint, HttpVersion.HTTP_1_1));
+    writer.println("Host: www.example.com");
+    writer.println("");
+    writer.flush();
+  }
+
+  private String getResponse(Socket socket) {
+    try (StringWriter writer = new StringWriter()) {
+      BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+      String line;
+      while (!isEmpty(line = reader.readLine())) {
+        writer.append(line).append("\r\n");
+      }
+      return writer.toString();
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  private void assertResponse(String response, boolean shouldBeValid) {
+    assertThat(!isEmpty(response), is(shouldBeValid));
+    if (shouldBeValid) {
+      assertThat(response, containsString("HTTP/1.1 200"));
+    }
+  }
+
+  private static class MuleContextStopper extends Thread {
```

Comment:
use an Executor instead of a thread (`Executors.single...`). Remember to shut it down on test tear down.